File: algorithms/javascript/validate-binary-search-tree/validate-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
 var isValidBST = function(root, flooring=Number.MIN_SAFE_INTEGER, ceiling=Number.MAX_SAFE_INTEGER) {
    if(!root) return true;
    if(root.val <= flooring || root.val >= ceiling) return false;
    return isValidBST(root.left, flooring, root.val) && isValidBST(root.right, root.val, ceiling);    
};

File: algorithms/go/validate-binary-search-tree/validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	const MAXINT = int(^uint(0) >> 1)
	const MININT = ^MAXINT
	return recValidBST(root, MININT, MAXINT)
}

func recValidBST(root *TreeNode, min, max int) bool {
	if root == nil {
		return true
	}
	if root.Val <= min || root.Val >= max {
		return false
	}
	return recValidBST(root.Left, min, root.Val) && recValidBST(root.Right, root.Val, max)
}